### ISSUE TITLE
Incrementals taking a long time, despite completing data

### ISSUE THREAD

**Issue by nigeloakford** [wamdam/backy2#87]
I am wondering if you can help with this issue - it appears to be something with backy2, but I cannot work out if it is a bug or intentional?  In your documentation you say that backy2 can backup a 600GB servers in half a minute, but although there is very little read/write traffic on our incrementals, the process itself seems to spend a lot of time following the backup doing something before exiting, but I don't understand what.

We are using Ceph Block storage as a source, a Ceph Object Gateway as the target (offsite, different cluster) and a local postgres service as the backy2 backend (running on SSDs).

I have run four test backups in total for this report (following multiple other backups on different servers, which showed this issue.)  These are the commands being run:

`2021-07-21 10:49:03,943 [257558] $ /usr/bin/backy2 backup -s 2021-07-21T10:49:02 -r /tmp/fileOs3izN rbd://pool/26331166-574d-40f4-be0f-4b3976785fee@2021-07-21T10:49:02 26331166-574d-40f4-be0f-4b3976785fee`

`2021-07-21 14:00:11,204 [264344] $ /usr/bin/backy2 backup -s 2021-07-21T13:59:36 -r /tmp/filemNd81j -f e27e7292-ea08-11eb-a285-cfc83c4284d3 rbd://pool/26331166-574d-40f4-be0f-4b3976785fee@2021-07-21T13:59:36 26331166-574d-40f4-be0f-4b3976785fee`

`2021-07-21 15:55:44,481 [266504] $ /usr/bin/backy2 backup -s 2021-07-21T15:55:41 -r /tmp/file444zjS -f 9571ae0e-ea23-11eb-a285-cfc83c4284d3 rbd://pool/26331166-574d-40f4-be0f-4b3976785fee@2021-07-21T15:55:41 26331166-574d-40f4-be0f-4b3976785fee`

`2021-07-22 10:08:30,603 [297936] $ /usr/bin/backy2 -v backup -s 2021-07-22T10:08:15 -r /tmp/fileQ7QcRH -f ba11dc38-ea33-11eb-a285-cfc83c4284d3 rbd://pool/26331166-574d-40f4-be0f-4b3976785fee@2021-07-22T10:08:15 26331166-574d-40f4-be0f-4b3976785fee`

This is the disk I am using for the backup testing:

```
user@server:~$ sudo rbd du pool/26331166-574d-40f4-be0f-4b3976785fee
NAME                                                      PROVISIONED  USED
26331166-574d-40f4-be0f-4b3976785fee@2021-07-22T10:08:15        1 TiB   85 GiB
26331166-574d-40f4-be0f-4b3976785fee                            1 TiB   85 GiB
```
 
As you can see, I am using a 1TB data disk with 85GB of data.  For the initial backup, the temp file of the difference is 1.2MB and it took 1hr 40mins to fully backup to our offsite location (1 Gbps link between sites). The backup started at 10:49, and between this time and 11:20 there was Ceph read data (400-500Mbps) and data transfer offsite (400Mbps). The rest of the backup time was just empty as far as network traffic was concerned.
 
Running the backup again with no changed data and doing an `rbd diff` without `--whole-object` created an empty diff array in the temp file, as expected, as there were no changes (the file just shows `[ ]`).  It still took 1hr 44mins to back up the server from this diff file.  There was a tiny spike ~3Mbps to our offsite location for the first 10 minutes, then nothing for the rest of the time, and just a 5 minute spike of ~3Mbps from Ceph at the start.
 
Running the backup again with no changed data and doing an `rbd diff` with `--whole-object` created a difference file of 1.2MB, the same size as the initial diff file, and the backup took 1hr 42mins.  I am aware of the Ceph bug (https://tracker.ceph.com/issues/50787) which is why I ran it with and without `--whole-object`.  There was ~15 minutes of traffic offsite, peaking at 3Mbps, and ~15 minutes of read traffic to Ceph, at 600-800Mbps.

backy.log for first 3 backups: https://pastebin.com/SaMyJEnY
backy stats for the first 3 backups: https://pastebin.com/tJX3pjMF

I ran a final test with the -v option (and without `--whole-object`).  There were a few changed blocks on this, although I believe that must have been Windows doing things, as I didn't change the data.  Anyway, diff file was only 2.2KB in size, so it wasn't much change.  The backup took 1hr 50mins.  There was ~20 minutes of <3Mbps traffic offsite, and less than 10 minutes of ~6Mbps traffic to Ceph.  The rest of the time showed no additional data transfer.

I have attached the full log to this post, as I can't upload it to pastebin due to the file size limit.  Please let me know if the temp file contents would be useful too.

[backup.log](https://github.com/wamdam/backy2/files/6862550/backup.log)

So basically my question is this - is it normal/expected behaviour that backy2 will spend over an hour (1:30hrs for some) processing backups after the data has been transferred?

Many thanks in advance.

**Comment by nigeloakford** [comment]
I have been doing some more testing on this, and it looks like the issue may be that the disks are underperforming hugely, so it is taking all this time to write into the database.  On a performance test I am getting ~0.5MB/s write speeds on the SSD, so each database transaction is taking a long time to complete.  I don't believe this is related to Backy itself after all!  I will be re-installing a blank Ubuntu install and checking the performance to try and narrow down what is causing the slow speeds.

**Comment by wamdam** [comment]
Yeah, sorry for the late reply. In backy2, sql performance has a huge impact. So having the sql database on an ssd or even nvme helps a lot.

**Comment by nigeloakford** [comment]
I am noticing that - even with SSDs it doesn't seem to keep up with it,  However, if I run multiple jobs, I can peak at 1700TPS on my database, but a single job can only max at about 100-120TPS even when it is running solo, which really seems to slow down the jobs.  Are there any changes I can make to backy or postgres that will speed up single job read/write speeds on the database?  It is still doing almost no network traffic after the initial burst, but has a 30-60 minute tail of database transactions before it completes.

**Comment by wamdam** [comment]
backy2 has a 2-step backup process. 1st step is to prepare the database, 2nd step is to backup data. The steps are separate so that backy2 can at any point continue a broken backup transaction-safe. So an initial "wait time" is expected.
